File: frontend/twinbar/src/utils/projectionCalculator.ts
/**
 * Statistical projection calculator for portfolio return forecasting
 *
 * Calculates 7-day future projections with confidence bands based on:
 * - Linear regression trend line
 * - Standard deviation for volatility
 * - Widening confidence intervals (fan out over time)
 */

import type { PriceDataPoint, ProjectionBand } from '../types/report';

/**
 * Calculate historical returns from price data
 * Returns cumulative return % from entry point for each day
 */
function calculateHistoricalReturns(data: PriceDataPoint[]): number[] {
  if (data.length === 0) return [];

  const returns: number[] = [];
  const entryPrice = data[0].close;

  for (const point of data) {
    // Calculate return % from entry point
    const returnPct = ((point.close - entryPrice) / entryPrice) * 100;
    returns.push(returnPct);
  }

  return returns;
}

/**
 * Calculate simple linear regression trend
 * Returns slope and intercept for y = mx + b
 */
function calculateLinearTrend(returns: number[]): {
  slope: number;
  intercept: number;
  predict: (x: number) => number;
} {
  const n = returns.length;
  if (n === 0) return { slope: 0, intercept: 0, predict: () => 0 };

  // Calculate means
  const xMean = (n - 1) / 2; // x values are 0, 1, 2, ..., n-1
  const yMean = returns.reduce((sum, y) => sum + y, 0) / n;

  // Calculate slope (m)
  let numerator = 0;
  let denominator = 0;
  for (let i = 0; i < n; i++) {
    const xDiff = i - xMean;
    const yDiff = returns[i] - yMean;
    numerator += xDiff * yDiff;
    denominator += xDiff * xDiff;
  }

  const slope = denominator !== 0 ? numerator / denominator : 0;
  const intercept = yMean - slope * xMean;

  return {
    slope,
    intercept,
    predict: (x: number) => slope * x + intercept,
  };
}

/**
 * Calculate standard deviation of returns (volatility measure)
 */
function calculateStandardDeviation(returns: number[]): number {
  if (returns.length === 0) return 0;

  const mean = returns.reduce((sum, val) => sum + val, 0) / returns.length;
  const squaredDiffs = returns.map((val) => Math.pow(val - mean, 2));
  const variance = squaredDiffs.reduce((sum, val) => sum + val, 0) / returns.length;

  return Math.sqrt(variance);
}

/**
 * Add days to a date
 */
function addDays(date: Date, days: number): Date {
  const result = new Date(date);
  result.setDate(result.getDate() + days);
  return result;
}

/**
 * Calculate 7-day future projections with statistical confidence bands
 *
 * @param historicalData - Historical price data points
 * @param initialInvestment - Starting portfolio value in dollars (e.g., 1000)
 * @param daysAhead - Number of days to project (default: 7)
 * @returns Array of projection bands with best/expected/worst case scenarios
 *
 * @example
 * ```typescript
 * const projections = calculateProjections(priceHistory, 1000, 7);
 * // Returns 7 projection bands, one for each future day
 * // projections[0] = tomorrow's projection
 * // projections[6] = 7 days from now projection
 * ```
 */
export function calculateProjections(
  historicalData: PriceDataPoint[],
  initialInvestment: number,
  daysAhead: number = 7
): ProjectionBand[] {
  if (historicalData.length === 0) {
    console.warn('calculateProjections: No historical data provided');
    return [];
  }

  // 1. Calculate historical returns from entry point
  const returns = calculateHistoricalReturns(historicalData);

  // 2. Calculate trend using linear regression
  const trend = calculateLinearTrend(returns);

  // 3. Calculate volatility (standard deviation of returns)
  const stdDev = calculateStandardDeviation(returns);

  // 4. Project future values with widening confidence bands
  const projections: ProjectionBand[] = [];
  const lastIndex = returns.length - 1;

  for (let day = 1; day <= daysAhead; day++) {
    const futureIndex = lastIndex + day;
    const futureDate = addDays(new Date(), day);

    // Expected return from trend line
    const expectedReturn = trend.predict(futureIndex);

    // Gradually widen confidence bands (fan out over time)
    // Day 1: 14% of std dev, Day 7: 100% of std dev
    const widening = day / daysAhead;

    // Best case: expected + std dev (widening)
    const bestCaseReturn = expectedReturn + (stdDev * widening);

    // Worst case: expected - std dev (widening)
    const worstCaseReturn = expectedReturn - (stdDev * widening);

    // Convert return % to portfolio NAV
    const expectedNav = initialInvestment * (1 + expectedReturn / 100);
    const bestCaseNav = initialInvestment * (1 + bestCaseReturn / 100);
    const worstCaseNav = initialInvestment * (1 + worstCaseReturn / 100);

    projections.push({
      date: futureDate.toISOString().split('T')[0],
      expected_return: Number(expectedReturn.toFixed(2)),
      best_case_return: Number(bestCaseReturn.toFixed(2)),
      worst_case_return: Number(worstCaseReturn.toFixed(2)),
      expected_nav: Number(expectedNav.toFixed(2)),
      best_case_nav: Number(bestCaseNav.toFixed(2)),
      worst_case_nav: Number(worstCaseNav.toFixed(2)),
    });
  }

  return projections;
}

/**
 * Enhance historical price data with portfolio metrics
 * Adds return_pct and portfolio_nav to each data point
 *
 * @param data - Historical price data
 * @param initialInvestment - Starting portfolio value
 * @returns Enhanced price data with portfolio metrics
 */
export function enhanceHistoricalData(
  data: PriceDataPoint[],
  initialInvestment: number
): PriceDataPoint[] {
  if (data.length === 0) return [];

  const entryPrice = data[0].close;

  return data.map((point) => {
    const returnPct = ((point.close - entryPrice) / entryPrice) * 100;
    const portfolioNav = initialInvestment * (1 + returnPct / 100);

    return {
      ...point,
      return_pct: Number(returnPct.toFixed(2)),
      portfolio_nav: Number(portfolioNav.toFixed(2)),
      is_projection: false,
    };
  });
}

File: src/analysis/projection_calculator.py
# -*- coding: utf-8 -*-
"""
Projection Calculator - Linear Regression Forecasting

Calculates 7-day forward projections with confidence bands based on historical price trends.

Ported from: frontend/twinbar/src/utils/projectionCalculator.ts

Algorithm:
1. Calculate historical returns from entry point
2. Fit linear trend line (slope × day_index + intercept)
3. Calculate volatility (standard deviation)
4. Project 7 days forward with widening confidence bands
"""
import logging
import numpy as np
from typing import List, Dict, Any
from datetime import datetime, timedelta

logger = logging.getLogger(__name__)


class ProjectionCalculator:
    """Calculate linear regression projections with confidence bands"""

    def __init__(self, initial_investment: float = 1000.0):
        """
        Initialize projection calculator

        Args:
            initial_investment: Starting portfolio value (default: $1000)
        """
        self.initial_investment = initial_investment

    def calculate_historical_returns(self, close_prices: List[float]) -> np.ndarray:
        """
        Calculate cumulative returns from first price (entry point)

        Args:
            close_prices: List of closing prices

        Returns:
            NumPy array of returns (%) from first price
        """
        if not close_prices:
            return np.array([])

        entry_price = close_prices[0]
        returns = []

        for price in close_prices:
            return_pct = ((price - entry_price) / entry_price) * 100
            returns.append(return_pct)

        return np.array(returns)

    def calculate_linear_trend(self, returns: np.ndarray) -> Dict[str, Any]:
        """
        Fit linear trend line to returns

        Args:
            returns: NumPy array of return percentages

        Returns:
            Dict with 'slope' and 'intercept'
        """
        if len(returns) == 0:
            return {'slope': 0.0, 'intercept': 0.0}

        # Edge case: single data point (can't fit line)
        if len(returns) == 1:
            return {'slope': 0.0, 'intercept': float(returns[0])}

        # Create x-axis: day indices
        x = np.arange(len(returns))

        # Use numpy polyfit for linear regression (degree 1)
        coefficients = np.polyfit(x, returns, 1)
        slope = coefficients[0]
        intercept = coefficients[1]

        return {
            'slope': float(slope),
            'intercept': float(intercept)
        }

    def calculate_standard_deviation(self, returns: np.ndarray) -> float:
        """
        Calculate standard deviation of returns (volatility measure)

        Args:
            returns: NumPy array of return percentages

        Returns:
            Standard deviation
        """
        if len(returns) == 0:
            return 0.0

        return float(np.std(returns))

    def calculate_projections(
        self,
        close_prices: List[float],
        dates: List[str],
        days_ahead: int = 7
    ) -> List[Dict[str, Any]]:
        """
        Calculate 7-day forward projections with confidence bands

        Algorithm:
        1. Calculate historical returns
        2. Fit linear trend
        3. Calculate volatility (std dev)
        4. Project forward with widening confidence bands

        Args:
            close_prices: List of historical closing prices
            dates: List of date strings (YYYY-MM-DD format)
            days_ahead: Number of days to project (default: 7)

        Returns:
            List of projection dicts with expected/best/worst case scenarios
        """
        if not close_prices or len(close_prices) == 0:
            return []

        # Calculate historical returns
        returns = self.calculate_historical_returns(close_prices)

        # Fit linear trend
        trend = self.calculate_linear_trend(returns)
        slope = trend['slope']
        intercept = trend['intercept']

        # Calculate volatility
        std_dev = self.calculate_standard_deviation(returns)

        # Last historical date
        # Handle both date strings and numeric indices
        last_date_str = dates[-1]

        # Check if it's a numeric string (e.g., "246" from RangeIndex)
        is_numeric = False
        if isinstance(last_date_str, (int, float)):
            is_numeric = True
        elif isinstance(last_date_str, str):
            try:
                float(last_date_str)
                is_numeric = True
            except ValueError:
                pass

        if is_numeric:
            # Numeric index - use today's date as fallback
            last_date = datetime.now()
            logger.warning(f"DataFrame has numeric index ({last_date_str}), using current date as projection start")
        else:
            last_date = datetime.strptime(str(last_date_str), '%Y-%m-%d')
        last_day_index = len(returns) - 1

        # Generate projections
        projections = []
        for future_day in range(1, days_ahead + 1):
            day_index = last_day_index + future_day
            projection_date = last_date + timedelta(days=future_day)

            # Expected return (linear trend)
            expected_return = slope * day_index + intercept

            # Confidence bands (widening factor based on how far into future)
            # Formula: day / days_ahead (day 7 = 100% widening, day 1 = 14% widening)
            widening_factor = future_day / days_ahead

            # Best/worst case = expected ± (std_dev × widening_factor)
            best_case_return = expected_return + (std_dev * widening_factor)
            worst_case_return = expected_return - (std_dev * widening_factor)

            # Convert returns to portfolio NAV
            expected_nav = self.initial_investment * (1 + expected_return / 100)
            best_case_nav = self.initial_investment * (1 + best_case_return / 100)
            worst_case_nav = self.initial_investment * (1 + worst_case_return / 100)

            projections.append({
                'date': projection_date.strftime('%Y-%m-%d'),
                'expected_return': round(expected_return, 2),
                'best_case_return': round(best_case_return, 2),
                'worst_case_return': round(worst_case_return, 2),
                'expected_nav': round(expected_nav, 2),
                'best_case_nav': round(best_case_nav, 2),
                'worst_case_nav': round(worst_case_nav, 2),
            })

        return projections

    def enhance_historical_data(self, ohlcv_data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """
        Add return_pct and portfolio_nav to historical OHLCV data

        Args:
            ohlcv_data: List of dicts with date, open, high, low, close, volume

        Returns:
            Enhanced list with return_pct and portfolio_nav fields
        """
        if not ohlcv_data:
            return []

        # Extract close prices
        close_prices = [point['close'] for point in ohlcv_data]

        # Calculate returns
        returns = self.calculate_historical_returns(close_prices)

        # Enhance data
        enhanced = []
        for i, point in enumerate(ohlcv_data):
            return_pct = returns[i] if i < len(returns) else 0.0
            portfolio_nav = self.initial_investment * (1 + return_pct / 100)

            enhanced_point = {
                **point,  # Preserve original fields
                'return_pct': round(return_pct, 2),
                'portfolio_nav': round(portfolio_nav, 2),
            }
            enhanced.append(enhanced_point)

        logger.info(f"✅ Enhanced {len(enhanced)} historical data points with portfolio metrics")
        return enhanced
